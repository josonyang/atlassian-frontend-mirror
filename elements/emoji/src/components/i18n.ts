import { defineMessages } from 'react-intl-next';

export const messages = defineMessages({
  deleteEmojiTooltip: {
    id: 'fabric.emoji.delete.tooltip',
    defaultMessage: 'Delete',
    description:
      'Tooltip content for delete emoji when focus on deletable emoji',
  },
  deleteEmojiTooltipForScreenreader: {
    id: 'fabric.emoji.delete.screenreader.tooltip',
    defaultMessage: 'To delete {shortName} emoji, press Backspace',
    description:
      'Tooltip content for delete emoji when focus on deletable emoji',
  },
  deleteEmojiTitle: {
    id: 'fabric.emoji.delete.title',
    defaultMessage: 'Remove emoji',
    description: 'Title for emoji removal dialog',
  },
  deleteEmojiDescription: {
    id: 'fabric.emoji.delete.description',
    defaultMessage:
      'All existing instances of this emoji will be replaced with {emojiShortName}',
    description: 'Description for emoji removal dialog',
  },
  deleteEmojiLabel: {
    id: 'fabric.emoji.delete.label',
    defaultMessage: 'Remove',
    description: 'Button label to remove emoji',
  },
  addCustomEmojiLabel: {
    id: 'fabric.emoji.add.custom.emoji.label',
    defaultMessage: 'Add your own emoji',
    description: 'Button label to add custom emoji',
  },
  emojiPlaceholder: {
    id: 'fabric.emoji.placeholder',
    defaultMessage: 'e.g. hello',
    description:
      'Placeholder for emoji that provides an example for emoji name',
  },
  emojiNameAriaLabel: {
    id: 'fabric.emoji.name.ariaLabel',
    defaultMessage: 'Enter a name for the new emoji',
    description: 'Explains to enter a name for a new emoji',
  },
  emojiChooseFileTitle: {
    id: 'fabric.emoji.choose.file.title',
    defaultMessage: 'Choose file',
    description: 'Choose custom emoji file',
  },
  emojiChooseFileScreenReaderDescription: {
    id: 'fabric.emoji.choose.file.screenReaderDescription',
    defaultMessage:
      'Choose a file for the emoji. JPG, PNG or GIF. Max size 1 MB',
    description:
      'Message indicating the purpose of choosing the file and requirements for the file',
  },
  emojiSelectSkinToneButtonAriaLabelText: {
    id: 'fabric.emoji.select.skin.tone.ariaLabel',
    defaultMessage: 'Choose your skin tone, {selectedTone} selected',
    description:
      'Message indicating the purpose of the skin tone selection button and the selected tone',
  },
  emojiSelectSkinToneListAriaLabelText: {
    id: 'fabric.emoji.select.skin.list.ariaLabel',
    defaultMessage: 'Skin tone selector',
    description:
      'Message indicating the purpose of the skin tone list and make user to choose one tone',
  },
  emojiImageRequirements: {
    id: 'fabric.emoji.image.requirements',
    defaultMessage: 'JPG, PNG or GIF. Max size 1 MB.',
    description: 'Message for emoji image requirements and maximum file size',
  },
  emojiPreviewTitle: {
    id: 'fabric.emoji.preview.title',
    defaultMessage: 'Preview',
    description: 'Emoji preview title',
  },
  emojiPreview: {
    id: 'fabric.emoji.preview',
    defaultMessage: 'Your new emoji {emoji} looks great',
    description: 'Emoji preview',
  },
  addEmojiLabel: {
    id: 'fabric.emoji.add.label',
    defaultMessage: 'Add emoji',
    description: 'verb - Button label to add emoji',
  },
  retryLabel: {
    id: 'fabric.emoji.retry.label',
    defaultMessage: 'Retry',
    description: 'verb - Button label to retry upload',
  },
  cancelLabel: {
    id: 'fabric.emoji.cancel.label',
    defaultMessage: 'Cancel',
    description: 'verb - button label to cancel operation',
  },
  searchPlaceholder: {
    id: 'fabric.emoji.search.placeholder',
    defaultMessage: 'Search',
    description: 'Placeholder for search emoji field',
  },
  searchLabel: {
    id: 'fabric.emoji.search.label',
    defaultMessage: 'Emoji name',
    description: 'verb - button label to search',
  },
  searchResultsStatus: {
    id: 'fabric.emoji.search.status',
    defaultMessage: 'Found {count} emojis',
    description: 'search results status for screenreader to read out',
  },
  searchResultsStatusSeeAll: {
    id: 'fabric.emoji.search.status',
    defaultMessage: 'Seeing all emojis',
    description:
      'search results status when no search query for screenreader to read out',
  },
  categoriesSelectorLabel: {
    id: 'fabric.emoji.categories.label',
    defaultMessage: 'Choose an emoji category',
    description:
      'Aria label for Emoji categories selector at the top of emoji picker',
  },
  categoriesSearchResults: {
    id: 'fabric.emoji.categories.search.results',
    defaultMessage: 'Search results',
    description: 'Emoji categories search results',
  },
  frequentCategory: {
    id: 'fabric.emoji.category.frequent',
    defaultMessage: 'Frequent',
    description: 'Emoji frequent category',
  },
  peopleCategory: {
    id: 'fabric.emoji.category.people',
    defaultMessage: 'People',
    description: 'Emoji frequent category',
  },
  natureCategory: {
    id: 'fabric.emoji.category.nature',
    defaultMessage: 'Nature',
    description: 'Emoji nature category',
  },
  foodsCategory: {
    id: 'fabric.emoji.category.foods',
    defaultMessage: 'Food & Drink',
    description: 'Emoji Foods category',
  },
  activityCategory: {
    id: 'fabric.emoji.category.activity',
    defaultMessage: 'Activity',
    description: 'Emoji activity category',
  },
  placesCategory: {
    id: 'fabric.emoji.category.places',
    defaultMessage: 'Travel & Places',
    description: 'Emoji Places category',
  },
  objectsCategory: {
    id: 'fabric.emoji.category.objects',
    defaultMessage: 'Objects',
    description: 'Emoji objects category',
  },
  symbolsCategory: {
    id: 'fabric.emoji.category.symbols',
    defaultMessage: 'Symbols',
    description: 'Emoji symbols category',
  },
  flagsCategory: {
    id: 'fabric.emoji.category.flags',
    defaultMessage: 'Flags',
    description: 'Emoji flags category',
  },
  productivityCategory: {
    id: 'fabric.emoji.category.productivity',
    defaultMessage: 'Productivity',
    description: 'Emoji productivity category',
  },
  userUploadsCustomCategory: {
    id: 'fabric.emoji.category.user.uploads',
    defaultMessage: 'Your uploads',
    description: 'User uploads in the custom category',
  },
  allUploadsCustomCategory: {
    id: 'fabric.emoji.category.all.uploads',
    defaultMessage: 'All uploads',
    description: 'All uploads in the custom category',
  },
  deleteEmojiFailed: {
    id: 'fabric.emoji.error.delete.failed',
    defaultMessage: 'Remove failed',
    description: 'Error message when custom emoji failed to be removed',
  },
  emojiInvalidImage: {
    id: 'fabric.emoji.error.invalid.image',
    defaultMessage: 'Selected image is invalid',
    description: 'Error message for invalid selected image',
  },
  emojiUploadFailed: {
    id: 'fabric.emoji.error.upload.failed',
    defaultMessage: 'Upload failed',
    description: 'Failed to upload emoji image',
  },
  emojiImageTooBig: {
    id: 'fabric.emoji.error.image.too.big',
    defaultMessage: 'Selected image is more than 1 MB',
    description: 'Error message for image too big, beyond the size limit',
  },
  emojiPickerTitle: {
    id: 'fabric.emoji.picker',
    defaultMessage: 'Emoji picker',
    description: 'Aria label for emoji picker',
  },
  emojiPickerListPanel: {
    id: 'fabric.emoji.pickerlist.tabpanel',
    defaultMessage: 'Emojis actions and list panel',
    description:
      'Aria lable for tabpanel of emoji picker, which shows emojis actions and list',
  },
  emojiPickerGrid: {
    id: 'fabric.emoji.pickerlist.grid',
    defaultMessage:
      '{showSearchResults, select, true{Search results} other{Emojis}}',
    description: `Aria label for emoji picker grid, showSearchResults is a boolean variable, message will be "Entering Emojis table", or "Leaving Emojis"`,
  },
  emojiButtonRoleDescription: {
    id: 'fabric.emoji.emojipicker.emoi.roledescription',
    defaultMessage: 'emoji button',
    description: `Aria roledescription for emoji button, used in emoji picker.`,
  },
  error: {
    id: 'fabric.emoji.emojipicker.error',
    defaultMessage: 'Error!',
    description: `Aria label for error icon, apperaed in emoji uploader screens and delete emoji screens of emoji picker`,
  },
});
