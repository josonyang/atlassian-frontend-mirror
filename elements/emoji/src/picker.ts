import { componentWithFG } from '@atlaskit/platform-feature-flags-react';

import { default as EmotionEmojiPicker } from './components/picker/EmojiPicker';
import { default as CompiledEmojiPicker } from './components/compiled/picker/EmojiPicker';
const EmojiPicker = componentWithFG(
	'platform_editor_css_migrate_emoji',
	CompiledEmojiPicker,
	EmotionEmojiPicker,
);
export { EmojiPicker };
