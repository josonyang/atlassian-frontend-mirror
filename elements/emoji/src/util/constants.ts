import type { CategoryId } from '../components/picker/categories';

export const customCategory = 'CUSTOM';
export const frequentCategory = 'FREQUENT';
export const customType = 'SITE';
export const searchCategory = 'SEARCH';
export const yourUploadsCategory = 'USER_CUSTOM';

export const customTitle = 'allUploadsCustomCategory';
export const userCustomTitle = 'userUploadsCustomCategory';

export const dataURLPrefix = 'data:';
export const deleteEmojiLabel = 'delete-emoji';

/**
 * A constant used in sorting/ordering to represent a number 'obviously much bigger than any item in the set being handled'.
 * This is used instead of Number.MAX_VALUE since subtraction of MAX_VALUE from itself occassionaly doesn't equal zero exactly :-(
 */
export const MAX_ORDINAL = 100000;

export const defaultEmojiHeight = 20;

export type EmojiPickerWidth = 350;
export const emojiPickerWidth: EmojiPickerWidth = 350;

export const sizeGap = 80;

export const defaultEmojiPickerSize = 'medium';

export const emojiPickerMinHeight = 260;
export const emojiPickerHeight = 295;
export const emojiPickerListHeight = emojiPickerHeight - 58; // picker height - actions height

export const emojiPickerPreviewHeight = 54;
export const emojiPickerHeightWithPreview = emojiPickerHeight + emojiPickerPreviewHeight;

export const localStoragePrefix = 'fabric.emoji';
export const selectedToneStorageKey = `${localStoragePrefix}.selectedTone`;
export const defaultCategories: CategoryId[] = [
	'PEOPLE',
	'NATURE',
	'FOODS',
	'ACTIVITY',
	'PLACES',
	'OBJECTS',
	'SYMBOLS',
	'FLAGS',
];

export enum KeyboardKeys {
	ArrowLeft = 'ArrowLeft',
	ArrowRight = 'ArrowRight',
	ArrowUp = 'ArrowUp',
	ArrowDown = 'ArrowDown',
	PageUp = 'PageUp',
	PageDown = 'PageDown',
	Home = 'Home',
	End = 'End',
	Enter = 'Enter',
	Tab = 'Tab',
	Space = ' ',
	Backspace = 'Backspace',
}

// Used to search available emoji in the picker list to focus
export enum KeyboardNavigationDirection {
	Down = 'Down',
	Up = 'Up',
	Left = 'Left',
	Right = 'Right',
}

export const CATEGORYSELECTOR_KEYBOARD_KEYS_SUPPORTED: string[] = [
	KeyboardKeys.ArrowRight,
	KeyboardKeys.ArrowLeft,
	KeyboardKeys.Home,
	KeyboardKeys.End,
];

export const EMOJI_KEYBOARD_KEYS_SUPPORTED: string[] = [
	KeyboardKeys.Enter,
	KeyboardKeys.Backspace,
	KeyboardKeys.Space,
];

export const TONESELECTOR_KEYBOARD_KEYS_SUPPORTED: string[] = [
	KeyboardKeys.Enter,
	KeyboardKeys.Tab,
	KeyboardKeys.Space,
];

export const EMOJIPICKERLIST_KEYBOARD_KEYS_SUPPORTED: string[] = [
	KeyboardKeys.ArrowRight,
	KeyboardKeys.ArrowLeft,
	KeyboardKeys.Home,
	KeyboardKeys.End,
	KeyboardKeys.ArrowUp,
	KeyboardKeys.ArrowDown,
	KeyboardKeys.PageUp,
	KeyboardKeys.PageDown,
];

export const DEFAULT_TONE = 0;

export const defaultListLimit = 50;

export const migrationUserId = 'hipchat_migration_emoticons';

export const analyticsEmojiPrefix = 'atlassian.fabric.emoji.picker';

export const EMOJI_LIST_COLUMNS = 8;

export const EMOJI_LIST_PAGE_COUNT = 5;

export const EMOJI_SEARCH_DEBOUNCE = 150;

// This is the base sampling rate in Emoji
export const SAMPLING_RATE_EMOJI_RENDERED_EXP = 20;
// This rate is used in fetching emoji resource
export const SAMPLING_RATE_EMOJI_RESOURCE_FETCHED_EXP = 100;
