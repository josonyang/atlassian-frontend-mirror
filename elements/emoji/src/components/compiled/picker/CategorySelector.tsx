/**
 * @jsxRuntime classic
 * @jsx jsx
 */
import { useCallback, useEffect, useRef, useState } from 'react';
import { css, jsx } from '@compiled/react';
import { cssMap, cx } from '@atlaskit/css';
import { token } from '@atlaskit/tokens';
import { N30 } from '@atlaskit/theme/colors';
import { useIntl } from 'react-intl-next';
import { Pressable } from '@atlaskit/primitives/compiled';
import Tooltip from '@atlaskit/tooltip';
import {
	CATEGORYSELECTOR_KEYBOARD_KEYS_SUPPORTED,
	defaultCategories,
	KeyboardKeys,
} from '../../../util/constants';
import type { CategoryDescription, OnCategory } from '../../../types';
import { messages } from '../../i18n';
import {
	CategoryDescriptionMap,
	type CategoryGroupKey,
	type CategoryId,
} from '../../picker/categories';
import { usePrevious } from '../../../hooks/usePrevious';
import { RENDER_EMOJI_PICKER_LIST_TESTID } from '../../picker/EmojiPickerList';

const styles = cssMap({
	commonCategory: {
		backgroundColor: token('color.background.neutral.subtle'),
		borderWidth: 0,
		borderRadius: token('border.radius'),
		paddingTop: token('space.0'),
		paddingBottom: token('space.0'),
		paddingLeft: token('space.0'),
		paddingRight: token('space.0'),
		transition: 'color 0.2s ease',
	},

	defaultCategory: {
		color: token('color.text.subtlest'),

		'&:hover': {
			color: token('color.text.selected'),
		},
	},

	activeCategory: {
		color: token('color.text.selected'),

		'&:hover': {
			color: token('color.text.selected'),
		},
	},

	disabledCategory: {
		color: token('color.text.subtlest'),
	},
});

const categorySelector = css({
	flex: '0 0 auto',
	backgroundColor: token('elevation.surface.sunken', N30),

	// eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
	button: {
		display: 'flex',
	},
});

const categorySelectorTablist = css({
	paddingTop: token('space.075', '6px'),
	paddingBottom: token('space.075', '6px'),
	paddingLeft: token('space.100', '8px'),
	paddingRight: token('space.100', '8px'),
	display: 'flex',
	flexDirection: 'row',
	justifyContent: 'space-around',
	alignItems: 'center',
});
export interface Props {
	dynamicCategories?: CategoryId[];
	activeCategoryId?: CategoryId | null;
	disableCategories?: boolean;
	onCategorySelected?: OnCategory;
}

export type CategoryMap = {
	[id: string]: CategoryDescription;
};

export const sortCategories = (c1: CategoryGroupKey, c2: CategoryGroupKey) =>
	CategoryDescriptionMap[c1].order - CategoryDescriptionMap[c2].order;

const addNewCategories = (
	oldCategories: CategoryId[],
	newCategories?: CategoryId[],
): CategoryId[] => {
	if (!newCategories) {
		return oldCategories;
	}
	return oldCategories
		.concat(newCategories.filter((category) => !!CategoryDescriptionMap[category]))
		.sort(sortCategories);
};

export const categorySelectorComponentTestId = 'category-selector-component';
export const categorySelectorCategoryTestId = (categoryId: string) =>
	`category-selector-${categoryId}`;

const CategorySelector = (props: Props) => {
	const { disableCategories, dynamicCategories, activeCategoryId, onCategorySelected } = props;
	const [categories, setCategories] = useState<CategoryId[]>(
		addNewCategories(defaultCategories, dynamicCategories),
	);
	const [currentFocus, setCurrentFocus] = useState(0);
	const categoryRef = useRef<HTMLDivElement>(null);
	const prevDynamicCategories = usePrevious(dynamicCategories);
	const { formatMessage } = useIntl();

	const updateCategories = useCallback(() => {
		const newCategories = addNewCategories(defaultCategories, dynamicCategories);
		setCategories(newCategories);
	}, [dynamicCategories]);

	useEffect(() => {
		if (prevDynamicCategories !== dynamicCategories) {
			updateCategories();
		}
	}, [prevDynamicCategories, dynamicCategories, updateCategories]);

	const focusCategory = useCallback(
		(index: number) => {
			const categoryToFocus: HTMLButtonElement | undefined | null =
				categoryRef.current?.querySelector(`[data-focus-index="${index}"]`);
			categoryToFocus && categoryToFocus.focus();
			setCurrentFocus(index);
		},
		[categoryRef, setCurrentFocus],
	);

	const handleKeyDown = (e: React.KeyboardEvent<HTMLButtonElement>) => {
		if (!CATEGORYSELECTOR_KEYBOARD_KEYS_SUPPORTED.includes(e.key)) {
			return;
		}
		e.preventDefault();
		e.stopPropagation();
		const lastCategoryIndex = categories.length - 1;
		switch (e.key) {
			// navigate to the right category
			case KeyboardKeys.ArrowRight:
				focusCategory(currentFocus === lastCategoryIndex ? 0 : currentFocus + 1);
				break;
			// navigate to the left category
			case KeyboardKeys.ArrowLeft:
				focusCategory(currentFocus === 0 ? lastCategoryIndex : currentFocus - 1);
				break;
			// navigate to the first category
			case KeyboardKeys.Home:
				focusCategory(0);
				break;
			// navigate to the last category
			case KeyboardKeys.End:
				focusCategory(lastCategoryIndex);
				break;
		}
	};

	const handleClick = (categoryId: CategoryId, index: number) => (event: React.SyntheticEvent) => {
		if (disableCategories) {
			event.preventDefault();
			return;
		}
		if (onCategorySelected) {
			onCategorySelected(categoryId);
		}
		setCurrentFocus(index);
	};

	let categoriesSection;
	if (categories) {
		categoriesSection = (
			<div
				role="tablist"
				aria-label={formatMessage(messages.categoriesSelectorLabel)}
				data-testid={categorySelectorComponentTestId}
				ref={categoryRef}
				css={categorySelectorTablist}
			>
				{categories.map((categoryId: CategoryId, index: number) => {
					const category = CategoryDescriptionMap[categoryId];

					const Icon = category.icon;
					const categoryName = formatMessage(messages[category.name]);
					return (
						<Tooltip content={categoryName} position="bottom" key={category.id}>
							<Pressable
								id={`category-selector-${category.id}`}
								data-focus-index={index}
								aria-label={categoryName}
								aria-controls={currentFocus === index ? RENDER_EMOJI_PICKER_LIST_TESTID : undefined}
								aria-selected={categoryId === activeCategoryId}
								xcss={cx(
									styles.commonCategory,
									styles.defaultCategory,
									categoryId === activeCategoryId && styles.activeCategory,
									disableCategories && styles.disabledCategory,
								)}
								isDisabled={disableCategories}
								onClick={handleClick(categoryId, index)}
								testId={categorySelectorCategoryTestId(categoryId)}
								tabIndex={currentFocus === index ? 0 : -1}
								onKeyDown={handleKeyDown}
								role="tab"
							>
								<Icon label={categoryName} />
							</Pressable>
						</Tooltip>
					);
				})}
			</div>
		);
	}
	return <div css={categorySelector}>{categoriesSection}</div>;
};

export default CategorySelector;
