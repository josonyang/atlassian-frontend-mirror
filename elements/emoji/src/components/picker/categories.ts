import BasketballIcon from '@atlaskit/icon/core/migration/basketball--emoji-activity';
import AddIcon from '@atlaskit/icon/core/migration/add--emoji-custom';
import FlagIcon from '@atlaskit/icon/core/migration/flag--emoji-flags';
import TakeoutFoodIcon from '@atlaskit/icon/core/migration/takeout-food--emoji-food';
import ClockIcon from '@atlaskit/icon/core/migration/clock--emoji-frequent';
import TreeIcon from '@atlaskit/icon/core/migration/tree--emoji-nature';
import LightbulbIcon from '@atlaskit/icon/core/migration/lightbulb--emoji-objects';
import EmojiIcon from '@atlaskit/icon/core/migration/emoji--emoji-people';
import HeartIcon from '@atlaskit/icon/core/migration/heart--emoji-symbols';
import VehicleCarIcon from '@atlaskit/icon/core/migration/vehicle-car--emoji-travel';
import CheckCircleIcon from '@atlaskit/icon/core/migration/check-circle--emoji-productivity';

import type { CategoryDescription } from '../../types';
import { customCategory, userCustomTitle, customTitle } from '../../util/constants';

export type CategoryId =
	| 'FREQUENT'
	| 'PEOPLE'
	| 'NATURE'
	| 'FOODS'
	| 'ACTIVITY'
	| 'PLACES'
	| 'OBJECTS'
	| 'SYMBOLS'
	| 'FLAGS'
	| 'ATLASSIAN'
	| 'CUSTOM';

export type CategoryGroupKey = CategoryId | 'USER_CUSTOM' | 'SEARCH';

type CategoryDescriptionRecord = {
	[key in CategoryGroupKey]: CategoryDescription;
};

export const CategoryDescriptionMap: CategoryDescriptionRecord = {
	SEARCH: {
		id: 'SEARCH',
		name: 'categoriesSearchResults', // refers to i18n categoriesSearchResults key
		icon: undefined,
		order: 0,
	},
	FREQUENT: {
		id: 'FREQUENT',
		name: 'frequentCategory',
		icon: ClockIcon,
		order: 1,
	},
	PEOPLE: {
		id: 'PEOPLE',
		name: 'peopleCategory',
		icon: EmojiIcon,
		order: 2,
	},
	NATURE: {
		id: 'NATURE',
		name: 'natureCategory',
		icon: TreeIcon,
		order: 3,
	},
	FOODS: {
		id: 'FOODS',
		name: 'foodsCategory',
		icon: TakeoutFoodIcon,
		order: 4,
	},
	ACTIVITY: {
		id: 'ACTIVITY',
		name: 'activityCategory',
		icon: BasketballIcon,
		order: 5,
	},
	PLACES: {
		id: 'PLACES',
		name: 'placesCategory',
		icon: VehicleCarIcon,
		order: 6,
	},
	OBJECTS: {
		id: 'OBJECTS',
		name: 'objectsCategory',
		icon: LightbulbIcon,
		order: 7,
	},
	SYMBOLS: {
		id: 'SYMBOLS',
		name: 'symbolsCategory',
		icon: HeartIcon,
		order: 8,
	},
	FLAGS: {
		id: 'FLAGS',
		name: 'flagsCategory',
		icon: FlagIcon,
		order: 9,
	},
	ATLASSIAN: {
		id: 'ATLASSIAN',
		name: 'productivityCategory',
		icon: CheckCircleIcon,
		order: 10,
	},
	USER_CUSTOM: {
		id: customCategory,
		name: userCustomTitle,
		icon: AddIcon,
		order: 11,
	},
	CUSTOM: {
		id: customCategory,
		name: customTitle,
		icon: AddIcon,
		order: 12,
	},
};
