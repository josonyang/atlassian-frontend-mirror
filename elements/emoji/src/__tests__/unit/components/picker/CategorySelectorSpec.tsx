import React from 'react';
import { messages } from '../../../../components/i18n';
import { CategoryDescriptionMap } from '../../../../components/picker/categories';
import CategorySelector, {
  type Props,
  sortCategories,
} from '../../../../components/picker/CategorySelector';
import { defaultCategories } from '../../../../util/constants';
import { isMessagesKey } from '../../../../util/type-helpers';
import type { CategoryId } from '../../../../types';
import { renderWithIntl } from '../../_testing-library';
import { fireEvent, screen } from '@testing-library/dom';
import { expectTabIndexFromList } from './_emoji-picker-test-helpers';

describe('<CategorySelector />', () => {
  const setupComponent = (props?: Props) =>
    renderWithIntl(<CategorySelector {...props} />);

  it('all standard categories visible by default', async () => {
    await setupComponent();
    const categoryButtons = await screen.getAllByRole('tab');
    expect(categoryButtons.length).toEqual(defaultCategories.length);
  });

  it('adds categories dynamically based on what has been passed in', async () => {
    await setupComponent({
      dynamicCategories: ['CUSTOM', 'FREQUENT'],
    });
    const categoryButtons = await screen.getAllByRole('tab');
    expect(categoryButtons.length).toEqual(defaultCategories.length + 2);
  });

  it('displays categories in sorted order', async () => {
    const dynamicCategories: CategoryId[] = ['CUSTOM', 'FREQUENT', 'ATLASSIAN'];
    await setupComponent({
      dynamicCategories,
    });
    const orderedCategories = dynamicCategories
      .concat(defaultCategories)
      .sort(sortCategories);
    const categoryButtons = await screen.getAllByRole('tab');
    orderedCategories.forEach((categoryId, i) => {
      const button: HTMLElement = categoryButtons[i];
      const categoryKey = CategoryDescriptionMap[categoryId].name;
      // eslint-disable-next-line
      expect(isMessagesKey(categoryKey)).toBeTruthy();
      if (isMessagesKey(categoryKey)) {
        expect(button).toHaveAttribute(
          'aria-label',
          messages[categoryKey].defaultMessage,
        );
      }
    });
  });

  it('all categories disabled if flag is set', async () => {
    await setupComponent({ disableCategories: true });
    const categoryButtons = await screen.getAllByRole('tab');
    expect(categoryButtons.length).toEqual(defaultCategories.length);
    defaultCategories.forEach((categoryId, i) => {
      const button = categoryButtons[i];
      const categoryKey = CategoryDescriptionMap[categoryId].name;
      expect(isMessagesKey(categoryKey)).toBeTruthy();
      if (isMessagesKey(categoryKey)) {
        expect(button).toHaveAttribute(
          'aria-label',
          messages[categoryKey].defaultMessage,
        );

        expect(button).toBeDisabled();
      }
    });
  });

  it('onCategorySelected called which clicking a category', async () => {
    let selectedCategoryId;
    await setupComponent({
      dynamicCategories: ['CUSTOM', 'FREQUENT'],
      onCategorySelected: (id) => {
        selectedCategoryId = id;
      },
    });
    const categoryButtons = await screen.getAllByRole('tab');
    const button = categoryButtons[defaultCategories.length + 1];
    fireEvent.click(button);
    expect(selectedCategoryId).toEqual('CUSTOM');
  });

  it('active category highlighted', async () => {
    const activeCategoryId = defaultCategories[3];
    await setupComponent({
      activeCategoryId,
    });
    const categoryButtons = await screen.getAllByRole('tab');
    // screen.debug()
    expect(categoryButtons.length).toEqual(defaultCategories.length);
    defaultCategories.forEach((categoryId, i) => {
      const button = categoryButtons[i];
      const categoryKey = CategoryDescriptionMap[categoryId].name;
      // eslint-disable-next-line
      expect(isMessagesKey(categoryKey)).toBeTruthy();
      if (isMessagesKey(categoryKey)) {
        expect(button).toHaveAttribute(
          'aria-label',
          messages[categoryKey].defaultMessage,
        );
      }
      const shouldBeActive = i === 3;
      if (shouldBeActive) {
        expect(button).toMatchSnapshot();
      }
    });
  });

  it('navigate categories via arrow keys', async () => {
    await setupComponent({
      dynamicCategories: ['CUSTOM', 'FREQUENT'],
    });
    const categoryButtons = await screen.getAllByRole('tab');
    const lastCategoryIndex = defaultCategories.length + 1;
    const customCategoryButton = categoryButtons[lastCategoryIndex];

    // first category should have tabIndex=0 by default
    expectTabIndexFromList(categoryButtons, 0);

    // focus on first category of categories
    categoryButtons[0].focus();
    expect(categoryButtons[0]).toHaveFocus();

    // press key left on first category, should move focus to last category
    fireEvent.keyDown(categoryButtons[0], { key: 'ArrowLeft' });
    expect(customCategoryButton).toHaveFocus();
    expectTabIndexFromList(categoryButtons, lastCategoryIndex);

    // press key right on last category, should move focus to first category
    fireEvent.keyDown(customCategoryButton, { key: 'ArrowRight' });
    expect(categoryButtons[0]).toHaveFocus();
    expectTabIndexFromList(categoryButtons, 0);

    // press key right on first category, should move focus to next one on the right hand side
    fireEvent.keyDown(categoryButtons[0], { key: 'ArrowRight' });
    expect(categoryButtons[1]).toHaveFocus();
    expectTabIndexFromList(categoryButtons, 1);

    // press key left on 2nd category, should move focus to next one on the left hand side
    fireEvent.keyDown(categoryButtons[1], { key: 'ArrowLeft' });
    expect(categoryButtons[0]).toHaveFocus();
    expectTabIndexFromList(categoryButtons, 0);

    // press key end should move focus to last category
    fireEvent.keyDown(categoryButtons[0], { key: 'End' });
    expect(customCategoryButton).toHaveFocus();
    expectTabIndexFromList(categoryButtons, lastCategoryIndex);

    // press key home should move focus to first category
    fireEvent.keyDown(customCategoryButton, { key: 'Home' });
    expect(categoryButtons[0]).toHaveFocus();
    expectTabIndexFromList(categoryButtons, 0);
  });
});
