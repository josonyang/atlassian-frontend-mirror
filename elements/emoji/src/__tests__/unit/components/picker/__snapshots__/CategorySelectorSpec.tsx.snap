// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategorySelector /> active category highlighted 1`] = `
.emotion-0 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  line-height: 1;
}

.emotion-0 >svg {
  overflow: hidden;
  pointer-events: none;
  max-width: 100%;
  max-height: 100%;
  color: var(--icon-primary-color);
  fill: var(--icon-secondary-color);
  vertical-align: bottom;
}

.emotion-0 >svg stop {
  stop-color: currentColor;
}

@media screen and (forced-colors: active) {
  .emotion-0 >svg {
    -webkit-filter: grayscale(1);
    filter: grayscale(1);
    --icon-primary-color: CanvasText;
    --icon-secondary-color: Canvas;
  }
}

.emotion-1 {
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  border: none;
  cursor: pointer;
  background-color: var(--ds-background-neutral-subtle, transparent);
  border: 0;
  border-radius: var(--ds-border-radius, 3px);
  padding: var(--ds-space-0, 0px);
  -webkit-transition: color 0.2s ease;
  transition: color 0.2s ease;
  color: var(--ds-text-subtlest, #7A869A);
  color: var(--ds-text-selected, #0052CC);
}

.emotion-1:focus,
.emotion-1:focus-visible {
  outline-color: var(--ds-border-focused, #2684FF);
  outline-offset: var(--ds-space-025, 2px);
  outline-style: solid;
  outline-width: var(--ds-border-width-outline, 2px);
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

@media screen and (forced-colors: active),screen and (-ms-high-contrast: active) {
  .emotion-1:focus-visible {
    outline: 1px solid;
  }
}

.emotion-1:hover {
  color: var(--ds-text-selected, #0052CC);
}

.emotion-1:hover {
  color: var(--ds-text-selected, #0052CC);
}

<button
  aria-label="Activity"
  aria-selected="true"
  class=" emotion-1"
  data-focus-index="3"
  data-testid="category-selector-ACTIVITY"
  id="category-selector-ACTIVITY"
  role="tab"
  tabindex="-1"
  type="button"
>
  <span
    aria-label="Activity"
    class="emotion-0"
    role="img"
    style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
  >
    <svg
      height="24"
      role="presentation"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M15.677 16.738A5.97 5.97 0 0112 18a5.97 5.97 0 01-3.677-1.262 6.329 6.329 0 001.982-4.589 6.33 6.33 0 00-2.157-4.746A5.973 5.973 0 0112 6c1.465 0 2.808.528 3.851 1.403a6.33 6.33 0 00-2.156 4.746c0 1.751.736 3.41 1.982 4.589zm1.354-1.472c.613-.94.969-2.063.969-3.266 0-1.124-.31-2.176-.85-3.076a4.321 4.321 0 00-1.455 3.225c0 1.187.495 2.313 1.336 3.117zm-10.062 0A5.964 5.964 0 016 12c0-1.124.31-2.176.85-3.076a4.324 4.324 0 011.455 3.225 4.323 4.323 0 01-1.336 3.117zM12 4a8 8 0 100 16 8 8 0 000-16z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
  </span>
</button>
`;
